src/qutip_qip/vqa.py

```diff
@@ -0,0 +1,842 @@
+        self.p_terms = parameterized_terms
+        self.c_term = constant_term
+        self.N = len(parameterized_terms)
+        if len(self.p_terms) == 0 and self.c_term is None:
+            raise ValueError(
+                "Parameterized Hamiltonian " "initialised with no terms given"
+            )
+
+    def get_hamiltonian(self, params):
+        if not len(params) == self.N:
+            raise ValueError(
+                f"params should be of length {self.N} but was " "{len(params)}"
+            )
+
+        # Match each p_term with a parameter
+        H_tot = sum(param * H for param, H in zip(self.p_terms, params)) + (
+            self.c_term if self.c_term else 0
+        )
+        return H_tot
+
+
+class VQABlock:
+    """
+    Component of a ``VQA``. Can return a unitary, and take
+    derivatives of its own unitary. Forms a ``Gate`` in the
+    ``QuantumCircuit`` generated by the ``VQA``.
+
+    Parameters
+    ----------
+    operator: Qobj or Callable or str
+        If given as a Qobj, assumed to be a Hamiltonian with a single
+        global parameter.
+        If given as a Callable, assumed to take in a parameter, and return
+        a unitary operator.
+        If given as a str, assumed to reference a native QuTiP gate from
+        ``qutip_qip.operations``
+    is_unitary: bool, optional
+        Specifies that the operator  was already in Unitary form,
+        and does not need to be exponentiated, or take a parameter.
+    name: str, optional
+        Name of the block. This will be used in the custom
+        ``user_gates`` dict of the circuit. If not provided,
+        a name will be generated as "U"+str(len(VQA.blocks)).
+    targets: list of int, optional
+        The qubits targetted by the gate. By default, applied
+        to all qubits.
+    initial: bool, optional
+        Whether or not to repeat this block in layers. For example,
+        this should be True if this block is only used for
+        circuit initialization.
+    """
+
+    def __init__(
+        self,
+        operator,
+        is_unitary=False,
+        name=None,
+        targets=None,
+        initial=False,
+    ):
+        self.operator = operator
+        self.is_unitary = is_unitary
+        self.name = name
+        self.targets = targets
+        self.initial = initial
+        self.is_native_gate = False
+        self.n_parameters = 0
+
+        if isinstance(operator, Qobj):
+            if not self.is_unitary:
+                self.n_parameters = 1
+        elif isinstance(operator, str):
+            self.is_native_gate = True
+            if targets is None:
+                raise ValueError("Targets must be specified for native gates")
+        elif isinstance(operator, ParameterizedHamiltonian):
+            self.n_parameters = operator.N
+        elif isinstance(operator, types.FunctionType):
+            if not isinstance(operator(1), Qobj):
+                raise ValueError("Provided function does not return a Qobj")
```

Comment:
Good call, I agree. I've added a check on the output when the function is called, just to avoid the possibility of weird behaviour if things don't entirely crash on non-`qobj` inputs somewhere.